fix: Use start_fid or page_token when fetching replicator messages

adityapk00 opened the pull request:
When paging through the replicator cursor, allow using the start_fid to start pages at that FID. This allows the client to break up and fetch multiple FIDs from multiple servers. Returned messages are in order (allowing them to be merged in order)

adityapk00 committed:
fix: Use start_fid or page_token when fetching replicator messages

sanjayprabhu commented on src/replication/replicator.rs:
```diff
@@ -411,7 +437,8 @@ impl Cursor {
 // cursor for all the various types of messages that need to be queried as part of the replication
-// process.
+// process. It will only fetch one type of messages `message_type`, and will break after the type
+// of messages have been exhausted
```
Where does this happen?

adityapk00 replied:
This is done in build_user_messages_for_fid. It uses the collect_messages() function with a specific message type, which will iterate over only the message postfix (so it won't go to the next message type)

sanjayprabhu commented on src/proto/replication.proto:
```diff
@@ -29,7 +29,10 @@ message GetShardSnapshotMetadataResponse {
-  optional bytes page_token = 3;
+  oneof cursor {
+    bytes page_token = 3;
+    uint64 start_fid = 4;
```
Is it worth passing in message types along with the fid?

adityapk00 replied:
Fixed.

adityapk00 committed:
Merge remote-tracking branch 'adityapk/fixfnameindex' into replicationsorting

adityapk00 committed:
send sort order over API

adityapk00 committed:
fix rare issue with multiple fnames per fid

adityapk00 committed:
misc: Log flaky test failure (#628)

Add extra logging to failing test so we can catch what is failing

adityapk00 committed:
batch size to 2k messages

sanjayprabhu committed:
fix: Add additional sync metrics to improve visibility (#631)

Track incoming and outgoing sync requests to help debug sync issues.

adityapk00 committed:
fix: Fix fname secondary DB index (#622)

The secondary index for fnames (To lookup an FID's UserNameProofs) is
missing for about ~1% of FIDs. It's not clear why this is missing, but
it's more common for older FIDs than newer ones, so probably this bug
existed in the code a while ago and is now fixed.

This PR adds a DB migration framework that can do both blocking (ie.
blocks engine startup) and background migrations. It adds the first
migration to fix the secondary index keys for UserNameProofs as a
background migration. It updates the DB version to schema version 2.

Read all the Fnames from the merkle trie, and check that they are
present in the DB and have the correct secondary index.

adityapk00 committed:
Merge remote-tracking branch 'origin/main' into replicationsorting

adityapk00 committed:
fix: Don't run parallel migrations + quiet migration logs (#632)

1. Don't log migration version checks too many times
2. Don't allow multiple migrations to run at once. - This is needed
because we construct the ShardEngine in many places, and we might
accidentally trigger multiple parallel migrations. While it might OK to
run them parallelly, we don't want to stress the DB too much.

adityapk00 committed:
Merge remote-tracking branch 'origin/main' into replicationsorting

adityapk00 committed:
fix: Supress version check log message (#633)

Supress the log message until we actually run the migrations

adityapk00 committed:
Merge remote-tracking branch 'origin/main' into replicationsorting

adityapk00 commented:
Closing this PR to make an identical one, merging into main
